### PR TITLE
ETH2 Validator node support

### PR DESCRIPTION
**Pull request by bachp**
###### Motivation for this change

**This is still work in progress, I'm currently testing and tweaking the modules: Don't us it yet as a mainnet validator. (Unless you know what you are doing).**.

Provide all the necessary services to allow running an [ETH2 validator](https://ethereum.org/en/eth2/staking/) on NixOS.

- [x] Services for ETH1 nodes
  - [x] [openethereum](https://github.com/openethereum/openethereum)
  - [x] [geth](https://github.com/ethereum/go-ethereum)
- [x] Services for ETH2 beacon and validator nodes
   - [x] [lighthouse](https://github.com/sigp/lighthouse)
- [ ] Package validator software
  - [ ] [lighthouse-ethereum](https://github.com/sigp/lighthouse)
  - [x] [lighthouse-ethereum-bin](https://github.com/sigp/lighthouse)

Open Points:

- [x] Package lighthouse from source. Currently the binary is packaged as the depndencies fo lighthouse can't be vendored. Fixed with v1.1.1. I will still provide an additional binary package as a fallback in case packaging issues arise again in the future.
- [ ] Create documentation how to setup a ETH2 validator
- [ ] lighthouse 1.1.1 build requires rust 1.50 https://github.com/NixOS/nixpkgs/pull/112792

Related MRs:

- [ ] #106983
- [ ] https://github.com/NixOS/nixpkgs/pull/113649

The goal is to have this all ready on unstable for the 21.05 stable release so that everybody can run a validatr node on stable.
This probably also means we should define that we will backport new releases of the involved packages to stable branches.

In addition to the official documentation a lot of the best practice was taken from the following sources:
- https://github.com/eth2-educators/eth2-docker
- https://someresat.medium.com/guide-to-staking-on-ethereum-2-0-ubuntu-lighthouse-41de20513b12

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    sha256 = "06yymky168av24w2xfpwpqiab74y90waiw88x92iq9dl6w800hlw";
+  };
+
+  # Work around the "unpacker appears to have produced no directories"
+  # case that happens when the archive doesn't have a subdirectory.
+  setSourceRoot = "sourceRoot=`pwd`";
+
+  nativeBuildInputs = [ autoPatchelfHook ];
+  buildInputs = [ stdenv.cc.cc.lib zlib ];
+  #buildPhase = "";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/lighthouse/default.nix:22]:
```suggestion
```

File: pkgs/applications/blockchains/openethereum/default.nix [hunk 2]
```diff
@@ -18,10 +19,17 @@ rustPlatform.buildRustPackage rec {
...
     repo = "openethereum";
     rev = "v${version}";
-    sha256 = "cs84Zz0nhagGDu5sDFTaFZF3SPEgJU8F4vGX7KLihOM=";
+    sha256 = "1qw4waifr5ziw82ly990y547g48mv9a0qv7f1q3ai1977mkkikvj";
   };
 
-  cargoSha256 = "6suNkHw1BbISb0MkYkUaD+mpUal+kn3y1SFVqzJFqJc=";
+  cargoPatches = [
+    (fetchpatch {
+      url = "https://github.com/openethereum/openethereum/commit/233bab2ee7c760327abbab85e5f82affea1a9f99.patch";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/openethereum/default.nix:27]: Please comment what this patch fixes.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 106063` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>lighthouse-ethereum</li>
<li>openethereum</li>
</ul>
</details>

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  version = "1.1.0";
+
+  src = fetchurl {
+    url = "https://github.com/sigp/lighthouse/releases/download/v${version}/lighthouse-v${version}-x86_64-unknown-linux-gnu-portable.tar.gz";
+    sha256 = "07vdhkj43fw6lbbwf5qzq0gamsg8zcgck78j74fj7irvxvkfidl7";
+  };
+
+  # Work around the "unpacker appears to have produced no directories"
+  # case that happens when the archive doesn't have a subdirectory.
+  setSourceRoot = "sourceRoot=`pwd`";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/lighthouse/default.nix:18]:
```suggestion
  setSourceRoot = "sourceRoot=$(pwd)";
```

File: pkgs/applications/blockchains/openethereum/default.nix [hunk 4]
```diff
@@ -38,7 +38,12 @@ rustPlatform.buildRustPackage rec {
   cargoBuildFlags = [ "--features final" ];
 
   # test result: FAILED. 88 passed; 13 failed; 0 ignored; 0 measured; 0 filtered out
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/openethereum/default.nix:40]:
```suggestion
```

File: pkgs/applications/blockchains/openethereum/default.nix [hunk 5]
```diff
@@ -38,7 +38,12 @@ rustPlatform.buildRustPackage rec {
...
-  doCheck = false;
+  #doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/openethereum/default.nix:41]:
```suggestion
```

File: pkgs/applications/blockchains/openethereum/default.nix [hunk 6]
```diff
@@ -38,7 +38,12 @@ rustPlatform.buildRustPackage rec {
...
+  # Fix tests by preventing them from writing to /homeless-shelter.
+  preCheck = "export HOME=$(mktemp -d)";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/openethereum/default.nix:43]:
```suggestion
  preCheck = ''
    export HOME=$(mktemp -d)
  '';
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 106063` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>openethereum: log was empty</li>
  </ul>
</details>
<details>
<summary>
1 package built:
</summary>
<ul>
<li>lighthouse-ethereum</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
lighthouse-ethereum:
<details>


`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/applications/blockchains/lighthouse/default.nix:24:3:

```
   |
24 |   installPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>
openethereum:
<details>


`gpl3` is a deprecated license, check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/blockchains/openethereum/default.nix:51:5:

```
   |
51 |     license = licenses.gpl3;
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
</details>
</details>

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  # Needed to get openssl-sys to use pkg-config.
+  OPENSSL_NO_VENDOR = 1;
+
+  # Prost build needs this
+  PROTOC = "${protobuf}/bin/protoc";
+  PROTOC_INCLUDE = "${protobuf}/include";
+
+  cargoSha256 = "0jjc40c8fabbcjp88yyix8b2g8b9rwx8v6h0kgban8hwk92d1yjp";
+
+  patches = [ ./0001-Allow-specifying-alternative-url-for-deposit_contrac.patch ./0002-Ignore-tests-that-require-network-access.patch ];
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/lighthouse/default.nix:52]:
```suggestion
  patches = [
    ./0001-Allow-specifying-alternative-url-for-deposit_contrac.patch
    ./0002-Ignore-tests-that-require-network-access.patch
  ];
```

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    sha256 = "151ahdj5114xb5hbhzsx5rx4yyry832hhm0bkx36s2np3r8v883q";
+  };
+
+  # LevelDB can't be unvendored and it requires cmake to be built
+  nativeBuildInputs = [ pkg-config cmake protobuf git ];
+  buildInputs = [ openssl zlib bzip2 leveldb ];
+  checkInputs = [ nodePackages.ganache-cli curl gnumake ];
+
+  # Needed to get openssl-sys to use pkg-config.
+  OPENSSL_NO_VENDOR = 1;
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/lighthouse/default.nix:44]: I would be happier if all the ENVs would be exported in preConfigure or where needed.

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    sha256 = "05irxnskcsdbp844hgk2hwkjmgfjxin99jr1fl5b5wg1v7v7gl4y";
+  };
+  eth2specs_unsafe = fetchFromGitHub {
+    owner = "sigp";
+    repo = "unsafe-eth2-deposit-contract";
+    # NOTE: the version of the unsafe contract lags the main tag, but the v0.9.2.1 code is compatible
+    # with the unmodified v0.12.1 contract. See `common/deposit_contract/build.rs` for details.
+    rev = "v0.9.2.1";
+    sha256 = "151ahdj5114xb5hbhzsx5rx4yyry832hhm0bkx36s2np3r8v883q";
+  };
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/lighthouse/default.nix:36]: This should be moved into a let in.

File: pkgs/top-level/all-packages.nix [hunk 10]
```diff
@@ -11597,7 +11597,7 @@ in
   }));
   spidermonkey_60 = callPackage ../development/interpreters/spidermonkey/60.nix { };
   spidermonkey_68 = callPackage ../development/interpreters/spidermonkey/68.nix { };
-  spidermonkey_78 = callPackage ../development/interpreters/spidermonkey/78.nix { };
+  spidermonkey_78 = callPackage ../development/interpreters/spidermonkey/78.nix { inherit (rustPackages_1_45) cargo rustc; };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:11600]:
```suggestion
  spidermonkey_78 = callPackage ../development/interpreters/spidermonkey/78.nix {
    inherit (rustPackages_1_45) cargo rustc;
  };
```

**Comment by SuperSandro2000** [comment]
If the rust update is intentional please move it into another PR.

**Comment by bachp** [comment]
> If the rust update is intentional please move it into another PR.

The rust update is required for lighthouse 1.1.3 to build. It's taken from https://github.com/NixOS/nixpkgs/pull/112792

I think at this point it makes no sense yet to review this PR in to much detail. I will probaly split up the different pieces into separate PRs.

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+in
+rustPlatform.buildRustPackage rec {
+  pname = "lighthouse";
+  version = "1.2.1";
+
+  src = fetchFromGitHub {
+    owner = "sigp";
+    repo = "lighthouse";
+    rev = "v${version}";
+    leaveDotGit = true; # Needed to get the proper version
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/lighthouse/default.nix:35]: Can we inject the version easily? Leaving dotGit is kind of a last resort.

Comment (bachp) [reply pkgs/applications/blockchains/lighthouse/default.nix:35]: I tried this with https://github.com/NixOS/nixpkgs/pull/106063/files#diff-d8165e975d221069554707389b93737b9aa76e9a4ac2b62e7f466bb8d3494bd5 but I need to find a better solution that I can submit upstream.

Leaving the dotGit doesn't even work for some reason. I will remove it in the next iteration.

**Comment by bachp** [comment]
Supreseeded by https://github.com/NixOS/nixpkgs/pull/128984

**Pull request closed without merging**